feature: add deepCopy & test code

craftedhaze92 opened the pull request:
- deepCopy 기능 구현
- Jest test code 추가

craftedhaze92 committed:
feature: add deepCopy & test code

- deepCopy 기능 구현
- Jest test code 추가

craftedhaze92 committed:
feature: add deepCopy & test code

- deepCopy 기능 구현
- Jest test code 추가

f-lab-dennis commented on deepCopy.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  // 순환 참조 방지
+  if (cache.has(value)) {
```
cache를 사용하지 않고 순환 참조 방지가 가능할까요?

f-lab-dennis commented on deepCopy.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  // Map 복사
+  if (value instanceof Map) {
```
혹시 WeakMap도 복사 가능할까요?

craftedhaze92 replied:
@f-lab-dennis WeakMap도 복사 시도 해보았는데요.. 이부분은 안되는것 같습니다..

f-lab-dennis commented on deepCopy.test.js:
```diff
@@ -0,0 +1,79 @@
+
+describe("deepCopy", () => {
+  // 기본 타입은 값 자체를 복사 하므로 원본과 복사본이 동일해야 한다.
+  test("Primitive 타입 복사", () => {
```
테스트 결과값을 확인하는 함수를 다양하게 잘 활용하신 것 같습니다. 그런데 하나의 test에 너무 많은 expect가 들어가 있어 만약 실패했을 경우 정확히 어느 primitive에서 실패했는지 파악하는 것이 직관적이지 않을 것 같습니다.

craftedhaze92 replied:
@f-lab-dennis primitive타입 테스트 케이스 상세하게 분리 하여 다시 반영하였습니다~!

f-lab-dennis commented on deepCopy.test.js:
```diff
@@ -0,0 +1,79 @@
+    const copy = deepCopy(arr);
+    expect(copy).toEqual(arr); // 배열 값이 동일함을 확인
+    expect(copy).not.toBe(arr); // 복사본이 원본과 동일한 참조가 아님을 확인
+    expect(copy[2]).not.toBe(arr[2]); // 중첩된 객체도 다른 참조임을 확인
```
실패 케이스도 포함해주신 부분 좋네요!

f-lab-dennis commented on deepCopy.js:
```diff
@@ -0,0 +1,60 @@
+
+  // Set 복사
+  if (value instanceof Set) {
+    const setCopy = new Set();
```
변수명은 되도록 명사형이면 좋을 것 같습니다.

craftedhaze92 replied:
@f-lab-dennis  변수명 명사형으로 수정하였습니다~!

f-lab-dennis reviewed (commented):
전반적으로 코드를 깔끔하게 작성해주셨고 테스트 코드도 익숙치 않은 것 치고는 잘 작성해주신 것 같습니다. 그런데 deepCopy 함수 내부에 if문이 다수 존재하는데 if문을 최대한 줄일 수 있는 방향으로 리팩토링해보는 것도 좋을 것 같습니다.

craftedhaze92 committed:
Merge branch 'feat/deep-copy' of https://github.com/DoHyunHwang/js-deep-copy into feat/deep-copy

craftedhaze92 committed:
Revert "feature: add deepCopy & test code"

This reverts commit b0475307f404e7720f9c4d577ee523a09225c27d.

craftedhaze92 committed:
feature: change deepCopy & test code

deepCopy.js 코드 수정
- cache 대신 Map 으로 대체
- 변수명 명사 형태로 변경

테스트 코드 수정
- Primitive 타입 별로 테스트 하도록 수정

craftedhaze92 committed:
feature: change deepCopy & test code

deepCopy.js 코드 수정
- cache 대신 Map 으로 대체
- 변수명 명사 형태로 변경

테스트 코드 수정
- Primitive 타입 별로 테스트 하도록 수정

craftedhaze92 committed:
Merge branch 'feat/deep-copy' of https://github.com/DoHyunHwang/js-deep-copy into feat/deep-copy

craftedhaze92 committed:
"feature: change deepCopy & test code"

deepCopy.js 코드 수정
- cache의 WeekMap 대신 Map으로 대체
- 변수명 명사 형태로 변경

테스트 코드 수정
- Primitive 타입 별로 테스트 하도록 수정

This reverts commit a21a41df11ed0df0a0a76ce96de004556c4a3ae2.

craftedhaze92 committed:
feature: 리팩토링 deep-copy.js

if 문을 최소화 하도록 코드 리팩토링

- 특정 타입들만 객체 형태로 정의하여 해당 객체를 호출하는 형태로 구현

f-lab-dennis commented on deepCopy.js:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  // 특정 타입 복사를 위한 객체
+  const constructors = {
```
재귀함수 내부에 선언한다면 상당히 많은 할당 연산이 발생할 것 같은데 외부로 빼는 것이 좋을 것 같습니다.

f-lab-dennis commented:
잘 구현해주셨습니다. 수고많으셨습니다~
